firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1035 @@
+numpy.import_array()
+
+include "dmplexinc.pxi"
+
+
+cdef inline void extents_reduce(void *in_, void *out, int *count, MPI.MPI_Datatype *datatype) nogil:
+    cdef:
+        PetscInt *xin = <PetscInt *>in_
+        PetscInt *xout = <PetscInt *>out
+
+    if xin[0] < xout[0]:
+        xout[0] = xin[0]
+    if xin[1] > xout[1]:
+        xout[1] = xin[1]
+    if xin[2] > xout[2]:
+        xout[2] = xin[2]
+    if xin[3] < xout[3]:
+        xout[3] = xin[3]
+
+
+@cython.wraparound(False)
+def layer_extents(mesh):
+    """
+    Compute the extents (start and stop layers) for an extruded mesh.
+
+    :arg mesh: The extruded mesh.
+
+    :returns: a numpy array of shape (npoints, 4) where npoints is the
+        number of mesh points in the base mesh.  ``npoints[p, 0:2]``
+        gives the start and stop layers for *allocation* for mesh
+        point ``p`` (in plex ordering), while ``npoints[p, 2:4]``
+        gives the start and stop layers for *iteration* over mesh
+        point ``p`` (in plex ordering).
+
+    .. warning::
+
+       The indexing of this array uses DMPlex point ordering, *not*
+       Firedrake ordering.  So you always need to iterate over plex
+       points and translate to Firedrake numbers if necessary.
+    """
+    cdef:
+        PETSc.DM dm
+        PETSc.SF sf
+        PETSc.Section section
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] cell_extents
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] tmp
+        PetscInt cStart, cEnd, c, cell, ci, p
+        PetscInt *closure = NULL
+        PetscInt closureSize
+        MPI.Datatype contig, typ
+        MPI.MPI_Op EXTENTS_REDUCER = NULL
+
+    dm = mesh._plex
+    section = mesh._cell_numbering
+    cell_extents = mesh.cell_set.layers_array
+    pStart, pEnd = dm.getChart()
+
+    iinfo = numpy.iinfo(IntType)
+
+    layer_extents = numpy.full((pEnd - pStart, 4),
+                               (iinfo.max, iinfo.min, iinfo.min, iinfo.max),
+                               dtype=IntType)
+    cStart, cEnd = dm.getHeightStratum(0)
+    for c in range(cStart, cEnd):
+        CHKERR(DMPlexGetTransitiveClosure(dm.dm, c, PETSC_TRUE, &closureSize, &closure))
+        CHKERR(PetscSectionGetOffset(section.sec, c, &cell))
+        for ci in range(closureSize):
+            p = closure[2*ci]
+            # Allocation bounds
+            # Each entity column is from bottom of lowest to top of highest
+            layer_extents[p, 0] = min(layer_extents[p, 0], cell_extents[cell, 0])
+            layer_extents[p, 1] = max(layer_extents[p, 1], cell_extents[cell, 1])
+            # Iteration bounds
+            # Each entity column is from top of lowest to bottom of highest
+            layer_extents[p, 2] = max(layer_extents[p, 2], cell_extents[cell, 0])
+            layer_extents[p, 3] = min(layer_extents[p, 3], cell_extents[cell, 1])
+    CHKERR(DMPlexRestoreTransitiveClosure(dm.dm, 0, PETSC_TRUE, NULL, &closure))
```

Comment:
So at this point `layers_extent[:, [0, 1]] == layers_extent[:, [2, 3]]`, but then you do a halo exchange on the latter.